### PR TITLE
fix(auth): handle InvalidStateException

### PR DESCRIPTION
**Pull request by Jordan-Nelson**
*Issue #, if available:* https://github.com/aws-amplify/amplify-flutter/issues/623

*Description of changes:*
- Create a new exception on the Flutter side, InvalidStateException
- On iOS, map AuthException.invalidState to InvalidStateException
- On Android, map FlutterInvalidStateException to InvalidStateException

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterInvalidStateException.kt [hunk 1]
```diff
@@ -18,5 +18,5 @@ package com.amazonaws.amplify.amplify_auth_cognito.types
 import com.amplifyframework.auth.AuthException
 
 class FlutterInvalidStateException: AuthException {
-    constructor(message: String, recoverySuggestion: String) : super(message, Exception("invalidSate"), recoverySuggestion)
+    constructor(message: String, recoverySuggestion: String) : super(message, Exception("invalidState"), recoverySuggestion)
```
Comment (cshfang) [inline packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterInvalidStateException.kt:21]: 🕵️   Is it possible that someone could be looking for `invalidSate` for some reason in their code?

File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterSignInRequest.kt [hunk 2]
```diff
@@ -55,7 +55,7 @@ data class FlutterSignInRequest(val map: HashMap<String, *>) {
       try {
         var user: AuthUser? = Amplify.Auth.currentUser;
         if (user != null) {
-          throw FlutterInvalidStateException("There is already a user  signed in.", "Sign out before calling sign in.")
+          throw FlutterInvalidStateException("There is already a user signed in.", "Sign out before calling sign in.")
```
Comment (cshfang) [inline packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterSignInRequest.kt:58]: 🕵️  Is it possible that someone could be looking for the previous string with the extra space for some reason in their code?

File: packages/amplify_auth_plugin_interface/lib/src/Exceptions/InvalidStateException.dart [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import 'package:amplify_auth_plugin_interface/amplify_auth_plugin_interface.dart';
+
+/// Thrown when an operation is not valid with the current state of Auth category 
+class InvalidStateException extends AuthException {
+  /// Named constructor
```
Comment (cshfang) [inline packages/amplify_auth_plugin_interface/lib/src/Exceptions/InvalidStateException.dart:20]: Nit: Don't think this is a named constructor
```suggestion
  /// Default constructor
```

File: packages/amplify_auth_cognito/android/src/test/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AmplifyAuthErrorHandlerTest.kt [hunk 4]
```diff
@@ -267,6 +268,30 @@ class AuthErrorHandlerTest {
         verify(mockResult, times(1)).error(expectedCode, defaultMessage, details);
     }
 
+    @Test
+    fun invalidStateException() {
+        // Arrange
+        val expectedCode = "InvalidStateException"
+        val exception: AuthException = FlutterInvalidStateException("Invalid state message", "Invalid state recovery message")
```
Comment (cshfang) [inline packages/amplify_auth_cognito/android/src/test/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AmplifyAuthErrorHandlerTest.kt:275]: Nit: Extract the strings for reuse

**Review by cshfang — APPROVED** [review]
Minor questions and a couple nits but lgtm otherwise 🚀

File: packages/amplify_auth_plugin_interface/lib/src/Exceptions/InvalidStateException.dart (hunk 3 above)
Comment (Jordan-Nelson) [reply packages/amplify_auth_plugin_interface/lib/src/Exceptions/InvalidStateException.dart]: You are right. This is what I get for copy and pasting code 🙂

File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterInvalidStateException.kt (hunk 1 above)
Comment (Jordan-Nelson) [reply packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterInvalidStateException.kt:21]: 🤯  ... Yes, it is definitely possible.

I am not sure if we would consider the message attached to an error part of the public API or not. It depends how you define public API (anything that is technically exposed as public or only the intended use).

Either way, this IMO is a bug fix and bug fixes should not be considered breaking in the same way that changing working functionality is considered breaking. Bug fixes are fixes to something that is currently broken. This entire PR is actually a good example of that. Currently the wrong exception is being thrown. This PR changes which exception type is being thrown. That could conceivably break code that a customer has written, but it is a fix to something that is broken and was never intended to be used that way.

**Review by dnys1 — APPROVED** [review]

**Review by haverchuck — APPROVED** [review]

**Pull request merged**